### ISSUE TITLE
WhatsApp closing the connection is not handled correctly

### ISSUE THREAD

**Issue by Half-Shot** [tulir/mautrix-whatsapp#21]
```
[Dec  1, 2018 23:26:52] [User/@Half-Shot:half-shot.uk/ERROR] WhatsApp error: unexpected websocket close: websocket: close 1000 (normal)
[Dec  1, 2018 23:26:53] [User/@Half-Shot:half-shot.uk/ERROR] WhatsApp error: unexpected websocket close: websocket: close 1000 (normal)
[Dec  1, 2018 23:26:54] [User/@Half-Shot:half-shot.uk/ERROR] WhatsApp error: unexpected websocket close: websocket: close 1000 (normal)
[Dec  1, 2018 23:26:55] [User/@Half-Shot:half-shot.uk/ERROR] WhatsApp error: unexpected websocket close: websocket: close 1000 (normal)
[Dec  1, 2018 23:26:56] [User/@Half-Shot:half-shot.uk/ERROR] WhatsApp error: unexpected websocket close: websocket: close 1000 (normal)
[Dec  1, 2018 23:26:57] [User/@Half-Shot:half-shot.uk/ERROR] WhatsApp error: unexpected websocket close: websocket: close 1000 (normal)
[Dec  1, 2018 23:26:58] [User/@Half-Shot:half-shot.uk/ERROR] WhatsApp error: unexpected websocket close: websocket: close 1000 (normal)
[Dec  1, 2018 23:26:59] [User/@Half-Shot:half-shot.uk/ERROR] WhatsApp error: unexpected websocket close: websocket: close 1000 (normal)
[Dec  1, 2018 23:27:00] [User/@Half-Shot:half-shot.uk/ERROR] WhatsApp error: unexpected websocket close: websocket: close 1000 (normal)
```

**Comment by rreuvekamp** [comment]
This seems to be a problem upstream: https://github.com/Rhymen/go-whatsapp/blob/c9a81e9578840e5c04c49c9e6126a1682faf22fc/conn.go#L276

There is an issue for it as well: https://github.com/Rhymen/go-whatsapp/issues/98

It seems that, at the moment, there is no way that the appservice can be informed of this connection problems, so we can't re-connect ourselves. So we'll have to fix this upstream.

**Comment by tulir** [comment]
Rhymen/go-whatsapp#34 is somewhat related: there's no way to disconnect cleanly.

**Comment by tulir** [comment]
This might be fixable now that Rhymen/go-whatsapp#114 added `Disconnect()`

**Comment by ara4n** [comment]
another similar failure mode:

```
error writing to socket: write tcp 192.168.0.128:42210->169.44.82.118:443: write: broken pipe
panic: repeated read on failed websocket connection

goroutine 42 [running]:
github.com/gorilla/websocket.(*Conn).NextReader(0xc00030e160, 0x116dee0, 0xc000441312, 0x2, 0xc000442040, 0xc00023cf00)
        /home/matrix/go/pkg/mod/github.com/gorilla/websocket@v1.4.0/conn.go:967 +0x356
github.com/gorilla/websocket.(*Conn).ReadMessage(0xc00030e160, 0xc002c520f0, 0xc000224f60, 0x1, 0x1, 0xb56700, 0xc002c520f0)
        /home/matrix/go/pkg/mod/github.com/gorilla/websocket@v1.4.0/conn.go:1028 +0x2f
github.com/Rhymen/go-whatsapp.(*Conn).readPump(0xc000306000)
        /home/matrix/go/pkg/mod/github.com/!rhymen/go-whatsapp@v0.0.0-20190208090600-c1173899de99/conn.go:272 +0x73
created by github.com/Rhymen/go-whatsapp.NewConn
        /home/matrix/go/pkg/mod/github.com/!rhymen/go-whatsapp@v0.0.0-20190208090600-c1173899de99/conn.go:129 +0x1ff
```

**Comment by tulir** [comment]
@ara4n That looks like about the same issue as @rreuvekamp had. That doesn't look like the latest version, so try updating since go-whatsapp had some major internal changes.

**Comment by tulir** [comment]
Should be fixed by 689c30407140f0943bb2a19644d7ce58d063c105
